src/metadata.rs

```diff
@@ -0,0 +1,109 @@
+use grpc_sys::{self, GrpcMetadataArray};
+use std::ffi::CString;
+use libc::c_char;
+#[allow(unused)]
+use std::ascii::AsciiExt;
+use std::slice;
+
+/// Builder used to construct a MetadataArray value.
+pub struct MetadataArrayBuilder {
+    entries: Vec<(*mut c_char, *mut c_char, usize)>
+}
+
+impl MetadataArrayBuilder {
+
+    /// Create a new empty builder.
+    pub fn new() -> MetadataArrayBuilder {
+        MetadataArrayBuilder { entries: vec![] }
+    }
+
+    /// Add a new key-value pair to the metadata being built.
+    pub fn add(mut self, key: Vec<u8>, value: Vec<u8>) -> MetadataArrayBuilder {
+        assert!(key.iter()
+            .all(|b|
+                     ((*b as u32 >= 0x30) && (*b as u32 <= 0x39))        // digits
+                         || ((*b as u32 >= 0x41) && (*b as u32 <= 0x5a)) // uppercase
+                         || ((*b as u32 >= 0x61) && (*b as u32 <= 0x7a)) // lowercase
+                         || (*b as u32 == 0x2e)                          // .
+                         || (*b as u32 == 0x2d)                          // -
+                         || (*b as u32 == 0x5f)                          // _
+            ));
+
+        let key_normalized = key.to_ascii_lowercase();
+
+        let value_size = value.len();
+        let pair = (
+            CString::new(key_normalized).unwrap().into_raw(),
+            CString::new(value).unwrap().into_raw(),
+            value_size
+        );
+        self.entries.push(pair);
+        self
+    }
+
+    /// Build the metadata array data type that can be used in CallOption
+    pub fn build(self) -> MetadataArray {
+        let array_size = self.entries.len();
+        let array = unsafe { grpc_sys::grpcwrap_metadata_array_create(array_size) };
+
+        for (key, value, value_size) in self.entries {
+            unsafe { grpc_sys::grpcwrap_metadata_array_add(array, key, value, value_size) };
+        }
+
+        MetadataArray { array }
+    }
+}
+
+/// Metadata data type used in CallOption data type.
+pub struct MetadataArray {
+    array: *mut GrpcMetadataArray,
+}
+
+impl MetadataArray {
+    pub fn as_mut_ptr(&mut self) -> *mut GrpcMetadataArray {
+        self.array
+    }
+}
+
+impl Drop for MetadataArray {
+    fn drop(&mut self) {
+        unsafe { grpc_sys::grpcwrap_metadata_array_destroy_full(self.array) }
+    }
+}
+
+/// Immutable handle to the core MetadataArray data type with accessor methods.
+pub struct MetadataArrayView {
```

Comment:
I think just `MetadataArray` is enough. You can make `as_mut_ptr` as `pub(crate)`.

Reply:
I agree, and this was just a rough first pass at it. It would be better to avoid having these two data types `MetadataArray` and `MetadataArrayView`.

However there might be a bit more to it than controlling visibility of as_mut_ptr.

Currently `MetadataArray` is used to handle the mutable GrpcMetadataArray used in `grpcwrap_call_X` calls. So it needs to be mutable because `grpcwrap_call_X` call `grpcwrap_metadata_array_move` which modifies the struct. However, the struct itself needs to be deallocated on the rust side (done via `Drop`.) 

In contrast `MetadataArrayView` is used for a `*const GrpcMetadataArray` returned from `grpcwrap_request_call_context_metadata`, and it is not supposed to be deallocated on the rust side, it is managed on the c side. Because of that, its lifetime should be dealt with appropriately so that it can't be used to reference the underlying memory that might be no longer invalid. That probably means that it should have the lifetime of corresponding `RequestContext` etc. but need to look into this.

But yes, it would be nice to unify these two ...

Reply:
Actually it's just the `Owned` and `Borrowed` form of the same struct/data. So you can pass `MetadataArray` to client call and return `&MetadataArray` from server call.

Reply:
Maybe we're not talking about the same thing? So who would own the `&MetadataArray` on the server call? Currently the server call provides `&MetadataArrayView` (with the same lifetime as the corresponding `RpcContext`) via the `RpcContext::metadata()` method. When the `MetadataArrayView` is deallocated (with the`RpcContext`) the underlying `*const GrpcMetadataArray` is *not deallocated*.

If `&MetadataArray` was used instead (as it is implemented now,) then the corresponding `*const GrpcMetadataArray` *would be deallocated*, which would be a double deallocate error, since on the server call this memory is all managed by grpc-core.

Reply:
> ... So who would own the &MetadataArray on the server call? ...

`RpcContext` will own the `MetadataArray` and `&MetadataArray` borrows it.

> If &MetadataArray was used instead (as it is implemented now,) then the corresponding *const GrpcMetadataArray would be deallocated, ...

No, it shouldn't. Because only a reference is returned. So just the reference is deallocated rather than the data it borrows.

I think the problem here is `GrpcMetadataArray` is defined as an `enum`. If it's declared as a C struct, then `*const GrpcMetadataArray` is free and safe to be cast to `&GrpcMetadataArray`.

Reply:
I have trouble understanding your previous suggestions. Perhaps you could take a stab at modifying the code to safely unify MetadataArray and MetadataArrayView? (You should be able to push into feature/metadata.)